### DISCUSSION TITLE
Typst book support?

### DISCUSSION THREAD

**Discussion by jimjam-slam** [quarto-dev/quarto-cli#6979] [Feature Requests]
### Description

Does anyone know if Quarto's new [Typst output format](https://quarto.org/docs/prerelease/1.4/typst.html) works with [book project types](https://quarto.org/docs/books/)? It isn't mentioned in the format list for books, and I haven't been able to convert my current book projects to use it, but I would love to see it built in at some point—seems like a perfect match!

(💡 I imagine you could jury-rig a single PDF from several docs with [includes](https://quarto.org/docs/authoring/includes.html)!)

**Comment by mcanouil** [comment]
For now, Typst is only for simple document and in the 1.4 pre-release.
Also it does not support yet all Quarto features, so that's something to add to the list.

**Reply by jimjam-slam** [reply]
No worries!

**Comment by rwaaijman** [comment]
just wondering if typst already support book projects?

**Reply by CarlColglazier** [reply]
It appears it does not as of version 1.5.43:

```
> quarto preview 

Preparing to preview
WARN: The typst format is not supported by book projects
ERROR: TypeError: Cannot read properties of undefined (reading 'pandoc')

Stack trace:
    at formatOutputFile (file:///Applications/quarto/bin/quarto.js:26760:29)
    at resolveInputTarget (file:///Applications/quarto/bin/quarto.js:74295:46)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async serveFiles (file:///Applications/quarto/bin/quarto.js:100432:24)
    at async serveProject (file:///Applications/quarto/bin/quarto.js:100056:30)
    at async Command.actionHandler (file:///Applications/quarto/bin/quarto.js:100756:9)
    at async Command.execute (file:///Applications/quarto/bin/quarto.js:8017:13)
    at async Command.parseCommand (file:///Applications/quarto/bin/quarto.js:7907:20)
    at async quarto (file:///Applications/quarto/bin/quarto.js:118093:9)
    at async file:///Applications/quarto/bin/quarto.js:118112:9
```

**Reply by mcanouil** [reply]
@CarlColglazier the question was about Typst supporting book not about Quarto supporting Typst book. If Typst does not have a Book template then Quarto is unlikely to have one. As far as I know there are no Typst book so far.

Once there is a book template then the team is very likely to work on it and use it to make Quarto Book project supports Typst.

**Reply by CarlColglazier** [reply]
Ah, I misunderstood!

**Reply by gordonwoodhull** [reply]
So far they have a fiction book template:

https://github.com/typst/templates/tree/main/wonderous-book

But nothing technical. So this is possible but we'd like to work with a template that is closer to our target.

**Reply by ratnanil** [reply]
Typst has many book and thesis templates available: https://typst.app/universe/search/?category=book%2Cthesis

**Reply by cscheid** [reply]
@ratnanil yes. But the difference is that those are in the Typst "universe"; if we make a book typst template, that will be Quarto's _default_. So we can't really choose any of those until one of them emerges as the community default for Typst.

**Reply by ratnanil** [reply]
yes that makes sense, thanks!

**Reply by roaldarbol** [reply]
> @ratnanil yes. But the difference is that those are in the Typst "universe"; if we make a book typst template, that will be Quarto's _default_. So we can't really choose any of those until one of them emerges as the community default for Typst.

Is this being tracked in an issue somewhere? What would it take for you to decide on one?

**Reply by mcanouil** [reply]
> Is this being tracked in an issue somewhere?

I don't think there is, at least not attached to this discussion.

> What would it take for you to decide on one?

Time mostly. And it's not about "decide on one", as explained by Carlos it's about making a Quarto one that fulfil all the Quarto requirements and that can be maintain by the team.

**Reply by roaldarbol** [reply]
Ah, no, I didn't mean "one" as in one template, I meant the requirements, the fields, etc. that would be needed. In the end, I guess it's a matter of making a decision-making; Typst has matured a lot over the past year when this post was first written, so maybe it's a conversation that would merit an issue. Then we could begin clarifying what the next steps would be and what the "right questions" are that need answering (and I know you're all busy, so I'm not expecting fast action, but a first step where y'all could pitch in).

**Reply by ratnanil** [reply]
The lack of book support is currently quite a drawback. I'm teaching students to use quarto and explicitly want to circumvent latex, but this also means missing out on quarto projects.

For me, it's not entirely clear if the quarto team want's to develop a book template, or whether you are waiting for an official typst book template. If it's the latter, I'm not sure the typst team is aware of this (I had a chat with them).

Btw, there is a book template maintained by the typst team ([wonderous-book](https://typst.app/universe/package/wonderous-book)).

**Reply by gordonwoodhull** [reply]
wonderous-book is a fiction book template.

I've been eyeing the Typst Orange Template, which looks more appropriate for technical publishing and seems to be actively maintained:

https://github.com/flavio20002/typst-orange-template

I can't promise when I will have time to try it, because we are a tiny team and we have a lot of other priorities, but it's something I'm strongly interested in.

It would be helpful to hear others' experience with the orange template or any others.

**Comment by MichelDSchmid** [comment]
I would like to take the liberty of providing a little input on this as well.
We use Quarto to create learning scripts for vocational students. Our aim is to have two versions (HTML and PDF), as identical as possible. Since these are technical professions, we use chapter numbering. That's why we use the "Book" project for Quarto. With Latex, however, we use the srcreprt document class, as in the example on the Quarto homepage. This would theoretically already be possible in typst ;-)
This only as an application purpose

regards
Mike

**Comment by DDyuti** [comment]
Hello! I want to render some python tables (PDF and HTML) using Typst+Quarto. Is there a way where I can apply styling /css to those tables and  render those designs perfectly on PDFs as well as HTML pages ?

**Reply by mcanouil** [reply]
@DDyuti your question is off topic. Please open a new GitHub discussion. Also, I suggest you take a look at the documentation about the two formats you mentioned.

**Comment by roaldarbol** [comment]
To take a slightly different approach to the question: What would it take to support a custom Typst book/thesis template to work with a Quarto book? Is it as simple as following [the custom Quarto/Typst template guide](https://quarto.org/docs/output-formats/typst-custom.html)? So if I've already got a Typst book template that I want to port to Quarto, do I just go ahead, or are there reasons it might not work? (I'm asking to avoid potentially wasting a couple of hours attempting to make it work in case it's futile or need some specific work on the Quarto end first)

**Reply by gordonwoodhull** [reply]
It’s not just a template, we also need some processing in the Quarto front-end to concatenate all the chapters into a book.

The way Quarto assembles a book, each chapter is rendered separately, and then the outputs are “concatenated” - but you can see that it’s not just pasting them together, the chapters have to be wrapped and put together in syntactically the right way.

If the performance of having all the chapters together in one document is okay for your project (e.g. including the chapters via shortcodes into a book document, not a book project), you could probably adapt a Typst book template today. But if you want the separate rendering of chapters, you’ll probably need to wait until it’s implemented in Quarto.

**Reply by roaldarbol** [reply]
Thanks for the reply! I might just make an attempt. I've made one for MyST Markdown, which I presume would just need to be adapted, but I'll see. Given the popularity of Typst, I'm really looking forward to Quarto supporting Typst books. :-)

**Reply by rainer-rq-koelle** [reply]
This would be a useful application for our use of R, Quarto, ... and now typst. @roaldarbol could you share your experience or template. I would love to make use of this.

**Reply by roaldarbol** [reply]
@rainer-rq-koelle I'm afraid I ended up exporting to LaTeX due to time-constraints. Maybe I'll find time eventually, but not soon.

**Reply by ratnanil** [reply]
I have a Typst template with book-like features (title pages, TOC, numbering, etc.) and I'm trying to understand the "adapt a Typst book template" approach @gordonwoodhull mentioned. What exactly does this mean? Could you outline this idea a bit more?

I've experimented with a book-merge-filter.lua that concatenates chapters before Typst processing, but Quarto still throws "format is not supported by book projects" errors. Is there a way to make Quarto recognize a custom Typst format as book-compatible?

**Reply by cscheid** [reply]
> Is there a way to make Quarto recognize a custom Typst format as book-compatible?

Not currently. A "book project" is a relatively special thing inside Quarto (because it needs to do, among other things, the kind of merging you're experimenting with yourself, and these are not really format-agnostic as currently implemented).

The implementation itself is at `src/project/types/book/book.ts`, but you'll see that it's really not setup well to support a new format.

**Reply by ratnanil** [reply]
Thank you very much for the clarification!

**Reply by rainer-rq-koelle** [reply]
Hi all, it has been a while. Was there any progress on this? Or could one direct me to a working example? 
Or do we have a set of other "work-arounds"? Thanks for any pointers ...

My team is keen to move to typst. 
Our "half-cooked" approach consists of "merging" the different quarto-chapters as `{{< include chapter-01.qmd >}}` etc. in the main document that has the typst format specifications (page-size, borders) in its YAML. 
We use a pagebreak-include after each chapter-include to have a new chapter starting on its own page.

"half-cook" features (we gave up finding a working solution for):

- we could not get the figure numbering to follow a chapter-number.figure-number notation. So figures are now numbered 1 through n.
- we have not experimented with cross-linking, but that should work
- we struggled a bit as our document requires moving from 1-column to 2-column sections. We place those ones into `:::{layout="[4,3]"} ... :::` divs. Most of these sections contain a figure (aka in the left column, widths 4 units) and surrounding text (aka in the right column, width 3 units).  There is some try-and-find-out to limit the text length (as longer text will result in whitespace below the figure) and move back to non-dived full page (1-column) layout. Precise viewers may spot a slight width difference between the div sections and the non-dived full page. P.S. do not use bullet point listings. These are somehow centred (or play with the text to make them "look" left aligned - we gave up on that one).

**Reply by gordonwoodhull** [reply]
Hi @rainer-rq-koelle. Your experience report is very helpful! 

Please note that you are responding to a comment from two weeks ago. There is no progress to report.

As @cscheid comments above, the real solution involves changing the Typescript that glues Quarto together. This would enable separate rendering of chapters.

What you are doing with the include short code is the best workaround that I know of. 

The next step would be to adapt a Typst template for books. This is not easy but it is likely to help with the figure numbering issue you mention, because it adds the chapter level of organization.

I have only seen two Typst templates for books. One is for fiction and is probably not relevant to most of our users. 

The other is this one, which looks more general and applicable to technical publishing:
https://github.com/flavio20002/typst-orange-template

Again it’s difficult work, but figuring out how well this template plays with Quarto is an important next step. It would be preferable to reuse the code without changing it, but that’s probably not possible. So this probably means forking the template and adapting it to add Quarto features from our default template.

I think that the column layout issue you mention is orthogonal to books, and probably more on the Typst side. I know that the Typst team is improving flow of text around images. (But I honestly don’t know if that work will help in your use case; this is a surprisingly big problem.)

**Reply by christopherkenny** [reply]
To the extent it is helpful, when writing my dissertation, I used the `{{< include file.qmd >}}` approach and it worked quite well.

This made the core file, `kenny-dissertation.qmd` look like this:

````
---
title: "Drawing Democracies: Redistricting in America"
format:
  harvard-diss-typst:
    keep-typ: false
... [the rest of the details on author, abstract, TOC, etc]
---

## Introduction {.introduction}

[A bunch of text directly in in the kenny-dissertation.qmd file.]

## Independence and Redistricting Commissions {#sec-letter}

{{< include 01-letter.qmd >}}

## Widespread Partisan Gerrymandering Mostly Cancels Nationally, but Reduces Electoral Competition {#sec-pbias}

{{< include 02-pbias.qmd >}}

## Principles and Partisanship: Explicit Rules Constrain Courts and Lawmakers {#sec-principles}

{{< include 03-principles.qmd >}}

## Redistricting Reforms Reduce Gerrymandering by Constraining Partisan Actors {#sec-didid}

{{< include 04-didid.qmd >}}

## Individual and Differential Harm in Redistricting {#sec-harm}

{{< include 05-harm.qmd >}}

## Appendix for Chapter 2 {.appendix}

{{< include 02-a-pbias.qmd >}}

## Appendix for Chapter 3

{{< include 03-a-principles.qmd >}}

## Appendix for Chapter 4

{{< include 04-a-didid.qmd >}}

## Appendix for Chapter 5

{{< include 05-a-harm.qmd >}}

````

The actual dissertation format, open-source at [christopherkenny/harvard-diss](https://github.com/christopherkenny/harvard-diss), then focused on the logic needed to style relevant pieces, including generating the frontmatter from the YAML. Overall, the approach works well and was fairly straightforward. A few specific features that I needed to meet the university guidelines:

1. The intro is numbered as Chapter 0 and appendices are A, B, ...: Handled with a [lua filter.](https://github.com/christopherkenny/harvard-diss/blob/main/_extensions/harvard-diss/special-sections.lua)
2. Figure and table numbers are shown as (Chapter #)-(Figure/Table #): The level 1 heading includes a counter update with its [show rule](https://github.com/christopherkenny/harvard-diss/blob/39d653242ecd099829a4b8d007d7729f492822f1/_extensions/harvard-diss/typst-template.typ#L73), which then works with a general [set rule](https://github.com/christopherkenny/harvard-diss/blob/39d653242ecd099829a4b8d007d7729f492822f1/_extensions/harvard-diss/typst-template.typ#L472) on the actual numbering.
3. I build a custom ToC which includes the title page, dedications, etc, before the normal chapters: This is done with an [organized set of queries](https://github.com/christopherkenny/harvard-diss/blob/39d653242ecd099829a4b8d007d7729f492822f1/_extensions/harvard-diss/typst-template.typ#L263)
4. Each chapter starts with the number in really big font on a new page: This is done by adding to the [same level 1 heading show rule ](https://github.com/christopherkenny/harvard-diss/blob/39d653242ecd099829a4b8d007d7729f492822f1/_extensions/harvard-diss/typst-template.typ#L76).
5. The last important thing was to ensure that headings in the included files were always adding using level 2 headings `##` at the highest, as this freed up level 1 headings `#` to act as chapters.

A copy of the dissertation is available [here](https://christophertkenny.com/drafts/kenny-dissertation.pdf), if you want to see what the format looks like.

**Comment by rwaaijman** [comment]
just wondering what the status is for support book projects within typst (with quarto)?

**Reply by mcanouil** [reply]
We appreciate your interest in the project, but I'd like to share some guidance on how we can make discussions more productive.

Posts asking "Any news?" or "When will this be ready?" without specific questions or constructive feedback don't help move development forward.
They create notification noise for maintainers without providing actionable input that could influence development priorities.

Rather than asking for general updates, consider asking specific technical questions about implementation, contributing implementation suggestions, or submitting pull requests for features you need.
Consider upvoting and/or adding a "thumb up" on the original message.

The most valuable contributions focus on "here's how I can help solve this" rather than "here's what I want you to do for me."

For the latest information about progress and planned features, check the project's changelog, release notes, or issues.

Thanks for understanding and for your continued interest in the project.

**Reply by ratnanil** [reply]
I'd be truly interested in knowing what I can do to help. I had a chat with the Typst developers, and I had the feeling it wasn't clear as to what the next step to supporting Typst books should be. Is there anything I can do to help?

**Reply by gordonwoodhull** [reply]
> Is there anything I can do to help?

Stay tuned. 😉

**Reply by statkclee** [reply]
+1

**Comment by cderv** [comment]
This feature request is now tracked in 
- https://github.com/quarto-dev/quarto-cli/issues/13954

**Comment by gordonwoodhull** [comment]
Typst Books are available in Quarto 1.9.18!

[orange-book](https://github.com/quarto-ext/orange-book) is bundled as the default Typst Book format extension. 

I also created Proof-of-Concept extensions for five other Typst Book packages. 

The first three worked well with Marginalia and the new Typst [article layout](https://prerelease.quarto.org/docs/authoring/article-layout.html) features. 

* [bookly](https://github.com/gordonwoodhull/quarto-bookly)
* [ilm](https://github.com/gordonwoodhull/quarto-ilm)
* [wonderous-book](https://github.com/gordonwoodhull/quarto-wonderous-book)

[haobook](https://github.com/gordonwoodhull/quarto-haobook) is already integrated with Marginalia. So it's "over-functioning" with many notes duplicated/overlapping. 

[min-book](https://github.com/gordonwoodhull/quarto-min-book) has large margins I couldn't figure out how to override, and it can glitch note locations or not resolve the layout. 

All Quarto book project features are fully functional on all five extensions (with tests!)

I do not intend to maintain these and I've archived them. Please fork and share your maintained version:
* Submit it to the [official Quarto extensions](https://github.com/quarto-dev/quarto-web/tree/main/docs/extensions/listings) listing
* Add it to the [unofficial Quarto extensions](https://m.canouil.dev/quarto-extensions/about.html) listing

Please open new discussions if you run into problems implementing a Quarto Typst book extension. These are the first book extensions in Quarto.
